### PR TITLE
lookupUnique

### PR DESCRIPTION
**Pull request by atzannes**
## Motivation and Context
I wanted a method that guarantees that if an object of a certain type exists, it is unique (there isn't a 2nd one). This can be useful for implementing primary-key semantics

## Approach
Implemented lookupUnique that returns None if there's no object, the object if it exists, and raises a TypeError if there is more than one objects. This works with indexing too, such that `Thing.lookupUnique(name="bla")` will raise if there's two Things named bla.

## Testing
Check out the unit-test I wrote

## Bonus
I was reading the LoginPlugin and AuthPlugin stuff and added some docstrings and used the ABC/@abstractmethod pattern where appropriate.

### CODE REVIEW

File: object_database/PyDatabaseObjectType.cpp [hunk 1]
```diff
@@ -1065,10 +1066,48 @@ PyObject* PyDatabaseObjectType::pyLookupAny(PyObject *databaseType, PyObject* ar
...
+
+        if (oid == NO_OBJECT) {
+            return incref(Py_None);
+        }
+        PyObject* result = obType->fromIntegerIdentity(oid);
+
+        // check uniqueness
+        oid = view->indexLookupNext(lookup.first, lookup.second, oid);
+        if (oid != NO_OBJECT) {
+            throw std::runtime_error(
```
Comment (braxtonmckee) [inline object_database/PyDatabaseObjectType.cpp:1100]: I think you may need to decref result here.

**Review by braxtonmckee — CHANGES_REQUESTED** [review]

File: object_database/PyDatabaseObjectType.cpp (hunk 1 above)
Comment (atzannes) [reply object_database/PyDatabaseObjectType.cpp:1100]: Good catch. Added the decref

Comment (atzannes) [reply object_database/PyDatabaseObjectType.cpp:1100]: Travis is happy. We're good to merge if you don't have additional comments.

**Comment by atzannes** [comment]
@braxtonmckee are we ready to merge this?

**Pull request merged**